### PR TITLE
[Core] refactor transformers 2d into multiple init variants.

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

For a new Transformer variant, we should do "Transformer2DModelForXXX" going forward. 

I would love an initial review of the design. 

Adjacent to https://github.com/huggingface/diffusers/pull/7489. I believe together with this PR and the current one, the class will read more modular.

### CODE REVIEW

File: src/diffusers/models/transformers/transformer_2d.py
Comment (DN6) [inline src/diffusers/models/transformers/transformer_2d.py:177]: I don't think all the args need to be passed here no? Think we can reduce to just these. They should be enough to configure the relevant attributes. 

```python
in_channels
norm_num_groups
use_linear_projection
sample_size
num_vector_embeds
inner_dim
interpolation_scale
norm_type
```

File: src/diffusers/models/transformers/transformer_2d.py
Comment (DN6) [inline src/diffusers/models/transformers/transformer_2d.py:262]: Since there's only a single block type used in Transformer2D at the moment,  I don't think we need a fetcher method just yet.

**Review by DN6 — COMMENTED** [review]
I like the init methods approach. Left a couple of comments.

**Comment by sayakpaul** [comment]
@DN6 reduced the call space quite a bit. See.

File: src/diffusers/models/transformers/transformer_2d.py
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_2d.py:176]:
```suggestion
```

File: src/diffusers/models/transformers/transformer_2d.py
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_2d.py:175]: do we need to pass any of the arguments? they are all accessible via `self.config`, no?

File: src/diffusers/models/transformers/transformer_2d.py
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_2d.py:173]: move these up, before calling the `self._init*()` methods
```suggestion
```

File: src/diffusers/models/transformers/transformer_2d.py
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_2d.py:152]: it seems like for continuous input, out_channnels should always be same as in_channels, not sure if we should add a check for it
```suggestion
        self.in_channels = in_channels
        self.out_channels = in_channels if out_channels is None else out_channels
        self.gradient_checkpointing = False
```

File: src/diffusers/models/transformers/transformer_2d.py
Comment (DN6) [inline src/diffusers/models/transformers/transformer_2d.py:187]: I don't think we need to define the transformer blocks inside each init method? Ideally the init methods should only configure attributes.

File: src/diffusers/models/transformers/transformer_2d.py
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_2d.py:211]:
```suggestion
```

File: src/diffusers/models/transformers/transformer_2d.py
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_2d.py:214]:
```suggestion
            self.proj_out = torch.nn.Linear(inner_dim, self.out_channels)
```

File: src/diffusers/models/transformers/transformer_2d.py
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_2d.py:216]:
```suggestion
            self.proj_out = torch.nn.Conv2d(inner_dim, self.out_channels, kernel_size=1, stride=1, padding=0)
```

File: src/diffusers/models/transformers/transformer_2d.py
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_2d.py:218]: same here, I think we can just use `self.config`
```suggestion
    def _init_vectorized_inputs(self):
```

File: src/diffusers/models/transformers/transformer_2d.py
Comment (DN6) [inline src/diffusers/models/transformers/transformer_2d.py:180]: Why are some arguments passed to the functions, while others are referenced from the config? IMO it's clearer pass the arguments to these init functions.

File: src/diffusers/models/transformers/transformer_2d.py
Comment (DN6) [inline src/diffusers/models/transformers/transformer_2d.py:177]: Can be moved to top level init and passed in as an argument to this function

File: src/diffusers/models/transformers/transformer_2d.py
Comment (DN6) [inline src/diffusers/models/transformers/transformer_2d.py:234]: This is the same for every init method right? It can be in the top level.

File: src/diffusers/models/transformers/transformer_2d.py
Comment (DN6) [inline src/diffusers/models/transformers/transformer_2d.py:337]: Not needed I think.

File: src/diffusers/models/transformers/transformer_2d.py
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_2d.py:318]:
```suggestion
            self.proj_out = nn.Linear(inner_dim, patch_size * patch_size * self.out_channels)
```

File: src/diffusers/models/transformers/transformer_2d.py
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_2d.py:337]: this is not used?

**Review by yiyixuxu — COMMENTED** [review]
very nice!! thanks for cleaning it up! much better now:)

File: src/diffusers/models/transformers/transformer_2d.py
Comment (sayakpaul) [reply src/diffusers/models/transformers/transformer_2d.py]: `norm_type` gets configured within init, so, it's safe to just pass `norm_type` and rely on `self.config` for the rest.

**Comment by sayakpaul** [comment]
@DN6 @yiyixuxu all the comments have been addressed. PTAL.

**Comment by sayakpaul** [comment]
@DN6 @yiyixuxu the problem with the failing tests (8) are a little shaky in nature. These failures happen ONLY when the input type is continuous. 

They're failing because the parameters of the `Transformer2Model` are changing depending on the position of its internal components (`transformer_blocks`, `proj_in`, etc.) EVEN WHEN seeded. 

Earlier, the `self.transformer_blocks` init was between `self.proj_in` and `self._proj_out`. This PR doesn't have that kind of positioning as the final part of the main `__init__()` initializes `self.transformer_blocks`. To confirm things, I put the init of `self.transformer_blocks` after 

https://github.com/huggingface/diffusers/blob/b990afd19cf53922cf943bc4623cb48bcd4f755d/src/diffusers/models/transformers/transformer_2d.py#L200

and before

https://github.com/huggingface/diffusers/blob/b990afd19cf53922cf943bc4623cb48bcd4f755d/src/diffusers/models/transformers/transformer_2d.py#L203

All the tests pass after this change. 

You might ask why does this not happen for the other input types (patched and vectorized)? Because we only configure `Transformer2DModel` for continuous input types in tests/models/test_layers_utils.py (which is where the 8 tests are failing). 

I see two potential solutions:

* Change the assertion values in the tests 
* Add `self.transformer_blocks` init in `_init_continuous_input()`.

LMK.

File: src/diffusers/models/transformers/transformer_2d.py
Comment (yiyixuxu) [reply src/diffusers/models/transformers/transformer_2d.py:170]: actually, these `init*` methods are not configuring attributes at all  - they mostly initialize the modules. I think in that case, it makes sense to define transformer blocks inside each in that case, even though it is a common model, the code will be duplicated 

e.g. this is 

```python
def _init_continuous_input(self):
        self.norm = torch.nn.GroupNorm(
            num_groups=self.config.norm_num_groups, num_channels=self.in_channels, eps=1e-6, affine=True
        )
        if self.use_linear_projection:
            self.proj_in = torch.nn.Linear(self.in_channels, self.inner_dim)
        else:
            self.proj_in = torch.nn.Conv2d(self.in_channels, self.inner_dim, kernel_size=1, stride=1, padding=0)

        # Use out_channels or default to in_channels
        if self.use_linear_projection:
            self.proj_out = torch.nn.Linear(self.inner_dim, self.out_channels)
        else:
            self.proj_out = torch.nn.Conv2d(self.inner_dim, self.out_channels, kernel_size=1, stride=1, padding=0)
```

**Comment by yiyixuxu** [comment]
>I see two potential solutions:

>Change the assertion values in the tests
>Add self.transformer_blocks init in _init_continuous_input() like the suggested above.

I left a comment here, https://github.com/huggingface/diffusers/pull/7491#discussion_r1544932116
I don't think we should move the transformer init inside these `_init*` methods just to get tests pass but I would actually in favor of doing that because it is easier to read IMO, depends on the input type, user can just go to one of the `init*` method instead of going back and forth; 
but I don't feel strongly about this, I think it is a nice clean up regardless. So I will leave it to you guys to decide :)

**Comment by sayakpaul** [comment]
> but I would actually in favor of doing that because it is easier to read IMO, depends on the input type, user can just go to one of the init* method instead of going back and forth;

This hits the point home. So, I will follow that. Thanks!

**Comment by yiyixuxu** [comment]
@sayakpaul @DN6 
should we actually just spin out `Transformer2DModelPatched`? we are pretty much already there - I think most of the work is already done

**Comment by sayakpaul** [comment]
Move out the classes to their own modules? I think it might be better done after https://github.com/huggingface/diffusers/pull/7489.

**Comment by yiyixuxu** [comment]
> Move out the classes to their own modules? I think it might be better done after https://github.com/huggingface/diffusers/pull/7489.
oh sure, 
let me know what you think @DN6: we are already creating 3 separate `__init__` and `forward`, they are practically already 3 different classes. I think the only models that use patch are pixart and vit; vectorized is only for vq diffusion but we will have to double-check

**Comment by sayakpaul** [comment]
I have an idea of the kind of refactoring @yiyixuxu suggesting. Would prefer doing it in a separate PR clubbing it with https://github.com/huggingface/diffusers/pull/7489. Would that be okay?

File: src/diffusers/models/transformers/transformer_2d.py
Comment (DN6) [reply src/diffusers/models/transformers/transformer_2d.py:181]: But the arguments to the transformer blocks are all the same right? None of these modules being created in the init methods are being passed to the transformer blocks? Why duplicate?

**Comment by DN6** [comment]
> @sayakpaul @DN6 should we just spin out `Transformer2DModelPatched` and `Transformer2DModelVectorized`? we are pretty much already there - I think most of the work is already done.

Yeah I'm cool with spinning it out into dedicated classes.  If possible, could they have model specific names e.g. PixartTransformer2D?

We can handle it in a separate PR if you prefer @sayakpaul 

Would we need to merge this PR if it's just going to be changed soonish? Can we just take the work done here and in https://github.com/huggingface/diffusers/pull/7489 and open a new PR?   

@yiyixuxu Is right that we are essentially creating three different init and forwards within the same object, so the signal here is that we should break it up into three different models.

**Comment by sayakpaul** [comment]
@DN6 

I would still prefer to handle the spinning out portion in https://github.com/huggingface/diffusers/pull/7489 and merge this PR. 

> If possible, could they have model specific names e.g. PixartTransformer2D?

I think we can gradually tackle these phasing out cases. More like a coarse-to-fine kind of approach. 

Regarding https://github.com/huggingface/diffusers/pull/7491#discussion_r1546340805, I think it's a matter of preference. So, I would like to elaborate a bit on why I am tending to go via the duplicating route for this particular case. 

The current PR allows the `Transformer2DModel` class to have the following structure:

```python
class Transformer2DModel:
    def __init__(self, ...):
        # init common attributes that are shared across the board
        self.in_channels = ...
        self.out_channels = ...

        # handle initialization of specific attributes, modules, etc.
        if input_type == "continuous":
            self._init_continuous_inputs()
        elif input_type == "patched":
            self._init_patched_inputs()
        elif input_type == "vectorized":
            self._init_vectorized_inputs()

    def _init_continuous_inputs(self):
        # initialize attributes specific to continuous inputs.
        ...

        # initialize blocks specific to continuous inputs.
        # input block.
        ...

        # transformer blocks.
        ... # <- duplicate code for other init methods too.

        # output blocks.
        ...

    # rest of the init methods follow the above structure too. 
```

Now, if we move out the transformer block initialization, it would look like:

```python
class Transformer2DModel:
    def __init__(self, ...):
        # init common attributes that are shared across the board
        self.in_channels = ...
        self.out_channels = ...

        # handle initialization of specific attributes, modules, etc.
        if input_type == "continuous":
            self._init_continuous_inputs()
        elif input_type == "patched":
            self._init_patched_inputs()
        elif input_type == "vectorized":
            self._init_vectorized_inputs()

        # transformer blocks.
        self.transformer_blocks = ...

    def _init_continuous_inputs(self):
        # initialize attributes specific to continuous inputs.
        ...

        # initialize blocks specific to continuous inputs.
        # input block.
        ...

        # output blocks.
        ...

    # rest of the init methods follow the above structure too. 
```

I think it breaks the linearity of the reading flow plus the tests (which are minor, I agree). 

Hope this provides more reasoning.

**Comment by sayakpaul** [comment]
@DN6 a gentle ping on the above ^.

**Review by DN6 — APPROVED** [review]
LGTM 👍🏽

**Pull request merged**
